image-builder/image_builder.sh

```diff
@@ -76,55 +76,83 @@ OK "init is installed"
 OK "Agent installed"
 [ "$(id -u)" -eq 0 ] || die "$0: must be run as root"
 
+ROOTFS_SIZE=$(du -B 1MB -s "${ROOTFS}" | awk '{print $1}')
 BLOCK_SIZE=${BLOCK_SIZE:-4096}
-IMG_SIZE=${IMG_SIZE:-80}
-
-info "Creating raw disk with size ${IMG_SIZE}M"
-qemu-img create -q -f raw "${IMAGE}" "${IMG_SIZE}M"
-OK "Image file created"
-
-# Kata runtime expect an image with just one partition
-# The partition is the rootfs content
-
-info "Creating partitions"
-parted "${IMAGE}" --script "mklabel gpt" \
-"mkpart ext4 1M -1M"
-OK "Partitions created"
 
-# Get the loop device bound to the image file (requires /dev mounted in the
-# image build system and root privileges)
-DEVICE=$(losetup -P -f --show "${IMAGE}")
+# Extra buffer space for the raw disk in percentage.
+IMG_BUFFER=30
 
-#Refresh partition table
-partprobe "${DEVICE}"
+# Calculate image size based on the rootfs
+calculate_img_size()
+{
+    IMG_SIZE=${IMG_SIZE:-$(((($ROOTFS_SIZE*$IMG_BUFFER)/100)+$ROOTFS_SIZE))}
+}
 
-MOUNT_DIR=$(mktemp -d osbuilder-mount-dir.XXXX)
-info "Formating Image using ext4 format"
-mkfs.ext4 -q -F -b "${BLOCK_SIZE}" "${DEVICE}p1"
-OK "Image formated"
+# Cleanup
+cleanup()
+{
+    sync
+    umount -l ${MOUNT_DIR}
+    fsck -D -y "${DEVICE}p1"
+    losetup -d "${DEVICE}"
+}
 
-info "Mounting root paratition"
-mount "${DEVICE}p1" "${MOUNT_DIR}"
-OK "root paratition mounted"
+create_rootfs()
+{
+    calculate_img_size
+    info "Creating raw disk with size ${IMG_SIZE}M"
+    qemu-img create -q -f raw "${IMAGE}" "${IMG_SIZE}M"
+    OK "Image file created"
+
+    # Kata runtime expect an image with just one partition
+    # The partition is the rootfs content
+
+    info "Creating partitions"
+    parted "${IMAGE}" --script "mklabel gpt" \
+    "mkpart ext4 1M -1M"
+    OK "Partitions created"
+
+    # Get the loop device bound to the image file (requires /dev mounted in the
+    # image build system and root privileges)
+    DEVICE=$(losetup -P -f --show "${IMAGE}")
+
+    #Refresh partition table
+    partprobe "${DEVICE}"
+
+    MOUNT_DIR=$(mktemp -d osbuilder-mount-dir.XXXX)
+    info "Formating Image using ext4 format"
+    mkfs.ext4 -q -F -b "${BLOCK_SIZE}" "${DEVICE}p1"
+    OK "Image formated"
+
+    info "Mounting root paratition"
+    mount "${DEVICE}p1" "${MOUNT_DIR}"
+    OK "root paratition mounted"
+    RESERVED_BLOCKS_PERCENTAGE=3
+    info "Set filesystem reserved blocks percentage to ${RESERVED_BLOCKS_PERCENTAGE}%"
+    tune2fs -m "${RESERVED_BLOCKS_PERCENTAGE}" "${DEVICE}p1"
+
+    AVAIL_DISK=$(df -B M --output=avail "${DEVICE}p1" | tail -1)
+    AVAIL_DISK=${AVAIL_DISK/M}
+    info "Free space root partition ${AVAIL_DISK} MB"
+
+    # if the available disk space is less than rootfs size, repeat the process
+    # of disk creation by adding 5% in the inital assumed value $ROOTFS_SIZE
+    if [ $ROOTFS_SIZE -gt $AVAIL_DISK ]; then
+        IMG_BUFFER=$(($IMG_BUFFER+5))
+        rm -f ${IMAGE}
+        unset IMG_SIZE
+        cleanup
+        create_rootfs
```

Comment:
Oooh - recursion ;)

I'd prefer to see a few more `info` calls to show:

- The attempt number

  "`Trying to create rootfs attempt $attempt`"-type message.

- The "old" and "new" images sizes the script is using

  Something like "`image size $old too small - trying again with size $new`".

Also, the final "`image created`" message is important. Currently it just says those words, but I think it would be good if it could contain useful information such as the final image size. This would allow consumers of this script (CI, tests, etc) to easily `grep` out the final value.. For example, we might want to use the auto-size feature to occasionally reset distro-specific image sizes and being able to automate that process would be valuable.

Reply:
Done. 

```
./image-builder/image_builder.sh -s 50  "${PWD}/fedora_rootfs"
Sending build context to Docker daemon  188.8MB
Step 1/2 : FROM fedora:latest
 ---> 422dc563ca32
Step 2/2 : RUN dnf install -y qemu-img parted gdisk e2fsprogs
 ---> Using cache
 ---> fa463f186699
Successfully built fa463f186699
Successfully tagged image-builder-osbuilder:latest
[OK] init is installed
[OK] Agent installed
INFO: Create root disk image. Attempt: 1
INFO: Creating raw disk with size 50M
[OK] Image file created
INFO: Creating partitions
[OK] Partitions created
INFO: Formating Image using ext4 format
[OK] Image formated
INFO: Mounting root paratition
[OK] root paratition mounted
INFO: Set filesystem reserved blocks percentage to 3%
tune2fs 1.43.5 (04-Aug-2017)
Setting reserved blocks percentage to 3% (368 blocks)
INFO: Free space root partition   41 MB
fsck from util-linux 2.30.2
e2fsck 1.43.5 (04-Aug-2017)
/dev/loop3p1: clean, 11/12288 files, 1423/12288 blocks
INFO: Create root disk image. Attempt: 2
INFO: Image size 50MB too small, trying again with size 179MB
INFO: Creating raw disk with size 179M
[OK] Image file created
INFO: Creating partitions
[OK] Partitions created
INFO: Formating Image using ext4 format
[OK] Image formated
INFO: Mounting root paratition
INFO: Set filesystem reserved blocks percentage to 3%
[OK] root paratition mounted
tune2fs 1.43.5 (04-Aug-2017)
Setting reserved blocks percentage to 3% (1359 blocks)
INFO: Free space root partition  147 MB
INFO: rootfs size 133 MB
INFO: Copying content from rootfs to root partition
[OK] rootfs copied
fsck from util-linux 2.30.2
e2fsck 1.43.5 (04-Aug-2017)
/dev/loop3p1: clean, 6586/45312 files, 38000/45312 blocks
INFO: Image created. Size: 179MB.
```